Adql2.1jakarta

pahjbo opened the pull request:
updates to be able to use with modern Java (>=9) where all the XML is now under the jakarta package

pahjbo committed:
initial maven publishing

slloyd-src committed:
- Changes from javax to jakarta

- updated to java17
- Minor changes due to java version changes

slloyd-src committed:
Updated starlink version

slloyd-src committed:
Removed unused code

slloyd-src committed:
-replaced tge 332 jar with 4.3 (version conflicts as one was controlled by gradle and the other in the libs folder) - NOTE: Gradle build needs fixing to remove the need for the STILS lib in the libs folder.

- Commented out the now seemingly defunct XTYPE usage (not available in 4.3)

slloyd-src committed:
Added a local XTYPE definition in the only place it references Starlink's definition (now removed)

slloyd-src committed:
updated starlink dependency

slloyd-src committed:
updated starling in ADQL

some modifications for debugging

slloyd-src committed:
- added publish to uksrc for all three libraries

- updated a couple of deprecated instantiation of class versions of simple types

pahjbo committed:
Merge pull request #1 from slloyd-src/jakarta

Jakarta updates for modern Java

slloyd-src committed:
Resolved javax/jakarta conflicts in test classes

Removed deprecated instantiation of Integer values

slloyd-src committed:
Removed deprecated instantiation of Double value

pahjbo committed:
Merge pull request #2 from slloyd-src/jakarta

Jakarta Pt2

slloyd-src committed:
UWS test updates

Updated instantiation of Integer in test classes
Added missing placeholders for the derived HttpServletRequest classes (in test only)

pahjbo committed:
Merge pull request #3 from slloyd-src/jakarta

UWS test updates

pahjbo committed:
remove uksrc repository specifics

gmantele commented:
This topic seems to be in the spirit of the times :smile: See #161 

You are perfectly right. As I said in the linked PullRequest, I have already performed this update in the branch [adql-2.1](https://github.com/gmantele/vollt/tree/adql2.1). This branch is still not stable ; TAP is not yet working in this branch, but I am currently working actively on it. I am doing my best so that it is ready by the end of this year, ideally the Autumn IVOA Interop. 

With this branch will come the following updates/changes:
- the dependency for the Jakarta package
- use exclusively Gradle (instead of Ant)
- new GitHub repository (see branch `adql-2.1` for a preview)

However, we have to stay with Java 8 because of TOPCAT currently limited to this version due to its user community (some users are still under Java 8). TOPCAT uses the ADQL-Library. I may see later for a solution using Gradle sub-projects regarding this issue.

pahjbo commented:
Note that these changes are onto the `adql-2.1` branch.

I have added my thoughts on trying to maintain Java 8 compatibility [here](https://github.com/gmantele/vollt/pull/161#issuecomment-2731964620) - basically do not try - it is too difficult! If the current version of the ADQLLib is published to maven central #148 then TOPCAT can depend on that until it wants to move beyond Java 8.

We will  be actively trying to use the UWS and TAP libraries from this branch, so might well be adding new PRs to fix things too.

bourgesl commented:
As I am an OpenJDK maintainer & share interest in long-term support on taplib/vollt ... for jmmc/eso,, I can make it for you = adopt multi-release jars (8,  11, 17, 21, 25).

We could talk about this task during VO days...

gmantele commented:
Thank you @pahjbo for your remarks. I agree with what you say, except for the way TOPCAT has to fetch VOLLT compiled in Java 8. Unless if I misunderstood, your solution means that TOPCAT cannot benefit from new features/fixes and ADQL language improvements because it will be stuck with an old version of VOLLT. My point is that it is fairly easy to change the Java version in `build.gradle` to the desired version. Going for a newer version is generally harmless, but not the contrary. That's why I try to keep the minimal version of Java to 8.

Ideally, it would be nice if I can do that only for the ADQL Library, but let the other libraries upgrade their Java version to 9+. I do not know whether it is possible, but I'd like to try few ideas before penalizing an entire community (TOPCAT and STILTS users). Until now, I keep this PullRequest open.

gmantele commented:
@pahjbo , @bourgesl , @vforchi , and others : there will be some VO events (in France, in the US, ...) this year. I would really like to have a chat together and think about the future of VOLLT. I am actually preparing myself for the OV France meeting next week and probably for the next interop as well. I have a lot of ideas and I think it is time to share all of them and see what good ideas we can have all together. I hope it is fine with you too.

Unfortunately, it is hard to make an important decision on several major topics in the same time (while minimizing backward compatibility issues). So, for the moment, I keep focusing entirely on finishing the support of ADQL-2.1 and TAP-1.1 in this branch `adql-2.1` (which makes already a huge amount of changes for ADQL and TAP libraries that you do not see yet in this branch because it does not compile yet). I also have a lot of ideas to improve the HTTP exchanges as well as the thread and job management and persistence for UWS and TAP, which I hope will improve compatibility with frameworks like Spring, Quarkus, .... while helping to design correct tests. But let's speak about all of that another time.... [I am already off-topic here :D]

mbtaylor commented:
@gmantele, thank you for your consideration for topcat users.  To keep entry and upgrade barriers as low as possible for topcat I would like to keep to a target platform of Java 8 while I can, and I've seen almost nothing in JDK 9+ that would benefit me or my users.  However I appreciate that VOLLT has its own constraints to consider, so if you decide you have to target a later Java version I'll work with it one way or another (stick with an earlier VOLLT version or bump my target platform).

I'm interested in @bourgesl suggestion of MRJARs, which I hadn't heard of before.  But I wonder if there's a more straightforward solution.  Could you use JDK 9 for the build, but (at least for the ADQL lib) cross-compile to java 8 classes (-source 8 -target 8)?  The UWS and TAP libraries still couldn't be used from Java 8 because of the missing `javax.servlet` classes, but I'm only interested in the ADQL library, which I'm guessing does not have a servlet dependency.

bourgesl commented:
Hi Mark,
To illustrate MR-JAR, here is an example I made to build JIDE-oss library on jdk8 / 9+ using the multi-release jar (special class implementations for 8 & 9), see this patch to fix maven build on 8/9: 
https://github.com/bourgesl/jide-oss/commit/3636dd0d0f91af0cb36eb20619d0ca8ea6bdf2a0

Cheers!

pahjbo commented:
> @gmantele, thank you for your consideration for topcat users. To keep entry and upgrade barriers as low as possible for topcat I would like to keep to a target platform of Java 8 while I can, and I've seen almost nothing in JDK 9+ that would benefit me or my users. However I appreciate that VOLLT has its own constraints to consider, so if you decide you have to target a later Java version I'll work with it one way or another (stick with an earlier VOLLT version or bump my target platform).

> I'm interested in @bourgesl suggestion of MRJARs, which I hadn't heard of before. But I wonder if there's a more straightforward solution. Could you use JDK 9 for the build, but (at least for the ADQL lib) cross-compile to java 8 classes (-source 8 -target 8)? The UWS and TAP libraries still couldn't be used from Java 8 because of the missing javax.servlet classes, but I'm only interested in the ADQL library, which I'm guessing does not have a servlet dependency.

That is probably the solution - the ADQL library should be in its own source repository (or the gradle build properly segregated) - it does not suffer from the jakarta issue that the UWS and TAP libraries do

tsdower commented:
At MAST, we have been using the vollt library by working from a fork with our own build scripts only for the ‘adql’ portion, which we wrap in a small microservice we’ve recently updated including the Jakarta changes. (This gets called by the rest of our TAP stack in Python.) As such I would like to note my support of suggestions for a separate ADQL build; it’s what we’ve been doing in a hacky way for some time.

slloyd-src committed:
Adjusted the lookup of tap.properties file slightly to allow the use of the default location if the custom location (isn't found).

Mainly for use in an IDE within an embedded app.
Previously default location would work but not if tapconf supplied but actual file not there.

slloyd-src committed:
Updated the home page to only use relative links to assist in deploying to more complicated proxy installations.

slloyd-src committed:
Updated the home page to only use relative links to assist in deploying to more complicated proxy installations.

slloyd-src committed:
changed to lowercase tap_schema for conformance with TAPSchemaLib and docs

slloyd-src committed:
Re-UPPERCASEd the TAP-SCHEMA name as that's what's being generated in the database by the archive-service and the tap_schema library at the moment

slloyd-src committed:
Added validation tests to check conformity with ADQL 2.1

Uses the ADQL Validator at https://github.com/ivoa/lyonetia/

slloyd-src committed:
added an env var to allow external unit tests to run if the VOLLT_BASE_PATH value is set.

slloyd-src committed:
Updated some ADQL Validator tests

Updated the default version of ADQL to 2.1
Added an env var VOLLT_BASE_PATH to the servlet path evaluation, in case of external usages other than standard usages where the context isn't defined (unit tests in parent projects).

slloyd-src committed:
added an env var to allow external unit tests to run if the VOLLT_BASE_PATH value is set.

slloyd-src committed:
Added fixes for windows line ending when comparing ADQL query strings in unit tests.

slloyd-src committed:
modified some unit test assertions to use platform-independent evaluations.

slloyd-src committed:
Forced the timezone to a specific timezone on a unit test so that running it in non-CET timezone works.

slloyd-src committed:
Changed a test path to one that SHOULD be available on Windows and Mac/Linux

slloyd-src committed:
-Windows CRLF removed IF present

- webAppRootDir HAS to be present as the internal logic requires it, so the java temp dir supplied for testing only.

slloyd-src committed:
Checks for URI before attempting to create a Path as file uris are interpreted as literal file paths.

slloyd-src committed:
Removed timezone comparison as comparing against a hard-coded date/time in a timezone other than Central Europe fails (local timezone isn't France)

slloyd-src committed:
Bracketed off a while loop correctly to stop the test from failing on linux

Removed a failing test of a deprecated method
Commented out a test method that fails due to the USW service (job summary) not returning the host/domain as expected

pahjbo committed:
add in the stil dependency

pahjbo committed:
added the uksrc repository config

slloyd-src committed:
Merge branch 'schema_rename' into jakarta

slloyd-src committed:
Merge remote-tracking branch 'origin/tapPropertiesLookup' into jakarta

slloyd-src committed:
Added some test logging

slloyd-src committed:
Merge branch 'ValidationTests' into adql2.1jakarta

slloyd-src committed:
Merge branch 'adql2.1jakarta' of https://github.com/uksrc/vollt into adql2.1jakarta

slloyd-src committed:
Added some test data

slloyd-src committed:
Merge branch 'adql2.1jakarta' into jakarta

slloyd-src committed:
Merge pull request #6 from slloyd-src/jakarta

Jakarta

slloyd-src committed:
Merge branch 'adql2.1jakarta' into deployment_fixes

# Conflicts:
#	ADQLLib/build.gradle
#	TAPLib/build.gradle

slloyd-src committed:
modified a token test for accented characters so that it would work on Windows too.

slloyd-src committed:
Updated gitignore for generated files (test)

slloyd-src committed:
exclude generated logs from git

slloyd-src committed:
Merge branch 'deployment_fixes' into adql2.1jakarta

# Conflicts:
#	.gitignore

pahjbo committed:
added the uksrc repository config
